monitor/src/main.rs

```diff
@@ -0,0 +1,289 @@
+            if response.is_none() {
+                tracing::warn!("failed to retrieve a block returned by the new block filter");
+                bail!("invalid block from block filter")
+            }
+            Ok(response.unwrap())
+        }
+    }))
+}
+
+struct Settlement {
+    time: DateTime<Utc>,
+    success: bool,
+}
+
+struct Count {
+    total: usize,
+    failures: usize,
+}
+
+struct SettlementFailureCounter {
+    // Note: settlements are not expected to be sorted by time because of uncle
+    // blocks.
+    settlements: Vec<Settlement>,
+    time_window: chrono::Duration,
+}
+impl SettlementFailureCounter {
+    fn new(time_window: Duration) -> Self {
+        SettlementFailureCounter {
+            settlements: Vec::new(),
+            time_window: chrono::Duration::from_std(time_window)
+                .expect("value too large for the chrono library"),
+        }
+    }
+
+    fn prune(&mut self) {
+        let threshold = Utc::now() - self.time_window;
+        self.settlements = std::mem::take(&mut self.settlements)
+            .into_iter()
+            .filter(|Settlement { time, .. }| time >= &threshold)
+            .collect();
+    }
+
+    fn push_settlement(&mut self, time: DateTime<Utc>, success: bool) {
+        self.settlements.push(Settlement { time, success })
+    }
+
+    fn settlements_from<'a>(
+        &'a self,
+        start: &'a DateTime<Utc>,
+    ) -> impl Iterator<Item = &'a Settlement> + 'a {
+        self.settlements
+            .iter()
+            .filter(move |Settlement { time, .. }| start <= time)
+    }
+
+    fn latest_count(&self) -> Count {
+        let mut total = 0;
+        let mut failures = 0;
+        for settlement in self.settlements_from(&(Utc::now() - self.time_window)) {
+            total += 1;
+            if !settlement.success {
+                failures += 1;
+            }
+        }
+        Count { total, failures }
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let args = Arguments::from_args();
+    shared::tracing::initialize(args.shared.log_filter.as_str());
+    tracing::info!("running monitoring service with {:#?}", args);
+
+    let transport = LoggingTransport::new(
+        web3::transports::Http::new(args.shared.node_url.as_str())
+            .expect("transport creation failed"),
+    );
+    let web3 = web3::Web3::new(transport);
+    let settlement_contract = GPv2Settlement::deployed(&web3)
+        .await
+        .expect("couldn't load deployed settlement");
+
+    let mut counter = SettlementFailureCounter::new(args.reverted_transaction_time_window);
+
+    let start =
+        Utc::now() - chrono::Duration::from_std(args.reverted_transaction_time_window).unwrap();
+    // Note: filtering out node error means that we might be ignoring failures.
+    let blocks = get_blocks_from(start, &web3)
+        .await
+        .filter_map(|x| async { x.ok() });
+
+    let mut txs = Box::pin(
+        blocks
```

Comment:
What's the difference between these three? I don't see much in the docs so I wouldn't know which to use when.

Reply:
pin_mut! is the most efficient because it does not allocate anything.
The reason Boxing helps in the first place is because the stream is `!Unpin` so it must be pinned before use and boxing makes it Unpin so it no longer has to be pinned.
Box::pin is the same thing as StreamExt::boxed.